design/plugin-backup-and-restore-progress-design.md

```diff
@@ -0,0 +1,448 @@
+Labels will be set with the value returned from `GetValidName` function. (https://github.com/vmware-tanzu/velero/blob/master/pkg/label/label.go#L35).
+
+If Plugin supports showing progress of the operation it is performing, it does following:
+- finds the VolumePluginBackup CR related to this backup operation by using `tags` passed in CreateSnapshot call
+- updates the CR with the progress regularly.
+
+After return from `CreateSnapshot` in `takePVSnapshot`, currently Velero adds `volume.Snapshot` to `backupRequest`. Instead of this, CR will be added to `backupRequest`.
+During persistBackup call, this CR also will be backed up to backup location.
+
+In backupSyncController, it checks for any VolumePluginBackup CRs that need to be synced from backup location, and syncs them to cluster if needed.
+
+VolumePluginBackup will be useful as long as backed up data is available at backup location. When the Backup is deleted either by manually or due to expiry, VolumePluginBackup also can be deleted.
+
+`processRequest` of `backupDeletionController` will perform deletion of VolumePluginBackup before volumesnapshotter's DeleteSnapshot is called.
+
+#### Backward compatibility:
+
+Currently `volume.Snapshot` is backed up as `<backupname>-volumesnapshots.json.gz` file in the backup location.
+
+As the VolumePluginBackup CR is backed up instead of `volume.Snapshot`, to provide backward compatibility, CR will be backed as the same file i.e., `<backupname>-volumesnapshots.json.gz` file in the backup location.
+
+For backward compatibility on restore side, consider below possible cases wrt Velero version on restore side and format of json.gz file at object location:
+
+- older version of Velero, older json.gz file (backupname-volumesnapshots.json.gz)
+
+- older version of Velero, newer json.gz file
+
+- newer version of Velero, older json.gz file
+
+- newer version of Velero, newer json.gz file
+
+First and last should be fine.
+
+For second case, decode in `GetBackupVolumeSnapshots` on the restore side should fill only required fields of older version and should work.
+
+For third case, after decode, metadata.name will be empty. `GetBackupVolumeSnapshots` decodes older json.gz into the CR which goes fine.
+It will be modified to return []VolumePluginBackupSpec, and the changes are done accordingly in its caller.
+
+If decode fails in second case during implementation, this CR need to be backed up to different file. And, for backward compatibility, newer code should check for old file existence, and follow older code if exists. If it doesn't exists, check for newer file and follow the newer code.
+
+`backupSyncController` on restore clusters gets the `<backupname>-volumesnapshots.json.gz` object from backup location and decodes it to in-memory VolumePluginBackup CR. If its `metadata.name` is populated, controller creates CR. Otherwise, it will not create the CR on the cluster. It can be even considered to create CR on the cluster.
```

Comment:
FYI: the backupsynccontroller doesn't directly contribute to restore.
But I get the idea.